### ISSUE TITLE
Regression when adding elements to lot of sequences

### ISSUE THREAD

**Issue by lscrd** [nim-lang/Nim#7299]
The following program runs without error on 0.17.2 and crashes with 0.18.0.

```nim
const N = 100_000_000
var a: array[1..N, seq[int]]

for i in 1..N: a[i] = @[1]
for i in 1..N: a[i].add(2)   # crashes here.
```

To see the problem, we have to allocate a lot of sequences (100 000 000 seems to be near the minimum). The problem can be seen in release mode or not. In the latter case, we see that the crash occurs in "alloc". Here is the stack trace:

```
add.nim(5)               add
gc.nim(586)              growObj
gc.nim(565)              growObj
alloc.nim(836)           rawDealloc
alloc.nim(528)           freeBigChunk
alloc.nim(483)           isAccessible
alloc.nim(331)           contains
alloc.nim(318)           intSetGet
SIGSEGV: Illegal storage access. (Attempt to read from nil?)
```

Changing the gc algorithm to "boehm" solves the problem in this case (but with "boehm", the actual program cannot execute successfully even in 0.17.2 as there are too many heap sections). Other algorithms cause a crash.

**Comment by Araq** [comment]
Works for me. Which OS? Maybe turn off memory overcommitment.

**Comment by lscrd** [comment]
I encountered the problem on Linux Manjaro 64 bits on a machine with 16GB of memory. Nim has been installed with *choosenim*. What do you mean by "turn off memory overcommitment"?

**Comment by mratsim** [comment]
Maybe it's linked to the new allocator? the 4GB limit per alloc/realloc maybe?

**Comment by zielmicha** [comment]
> What do you mean by "turn off memory overcommitment"?
`sudo sysctl vm.overcommit_memory=1`. This disables controversial Linux kernel feature that uses heuristics to prevent over-commitment of memory. This setting is recommended e.g. when running Redis.

**Comment by Araq** [comment]
>  the 4GB limit per alloc/realloc maybe?

That limit does not exist in 0.18.

**Comment by lscrd** [comment]
@zielmicha
Thanks! I didn't know about that. Value 0 seems indeed somewhat questionable.

I have tried with value 1 with the same result. I have also tried on the same machine booting on Ubuntu 17.10. The same executable (previously compiled on Manjaro) crashes with the same error. Recompiling on Ubuntu doesn't change anything.

The program consumes a lot of memory but when it crashes there is still more than 7 GB available. The version compiled with 0.17.2, which runs successfully, almost consumes all the available memory, but without swapping.
